src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,760 @@
+  keepAliveTimeout: number = 5_000;
+  port: number | null = null;
+
+  constructor(options?: ServerOptions, requestListener?: RequestListener) {
+    if (!flags.enableNodejsHttpServerModules) {
+      throw new ERR_METHOD_NOT_IMPLEMENTED('Server');
+    }
+    super();
+
+    if (options != null) {
+      // @ts-expect-error TS2345 TODO(soon): Find a better way to handle this type mismatch.
+      storeHTTPOptions.call(this, options);
+    }
+
+    // TODO(soon): Support options.highWaterMark option.
+
+    if (typeof options === 'function') {
+      requestListener = options;
+      options = {};
+    } else if (options == null) {
+      options = {};
+    } else {
+      validateObject(options, 'options');
+    }
+
+    if (requestListener) {
+      this.on('request', requestListener);
+    }
+
+    this[kUniqueHeaders] = parseUniqueHeadersOption(
+      options.uniqueHeaders as (string | string[])[]
+    );
+  }
+
+  close(callback?: VoidFunction): this {
+    httpServerPreClose(this);
+    if (this.port != null) {
+      portMapper.delete(this.port);
+      serverRegistry?.unregister(this);
+      this.port = null;
+    }
+    this.emit('close');
+    queueMicrotask(() => {
+      callback?.();
+    });
+    return this;
+  }
+
+  closeAllConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  closeIdleConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  setTimeout(
+    msecs?: number | ((socket: Socket) => void),
+    callback?: (socket: Socket) => void
+  ): this {
+    if (typeof msecs === 'function') {
+      callback = msecs;
+      msecs = undefined;
+    } else if (typeof msecs === 'number') {
+      this.timeout = msecs;
+    }
+
+    if (typeof callback === 'function') {
+      this.once('timeout', callback);
+    }
+    return this;
+  }
+
+  async #onRequest(request: Request): Promise<Response> {
+    const { incoming, response } = this.#toReqRes(request);
+    this.emit('connection', this, incoming);
+    this.emit('request', incoming, response);
+    return getServerResponseFetchResponse(response);
+  }
+
+  #toReqRes(request: Request): {
+    incoming: IncomingMessage;
+    response: ServerResponse;
+  } {
+    const incoming = new this[kIncomingMessage]();
+    const reqUrl = new URL(request.url);
+    incoming.url = reqUrl.pathname + reqUrl.search;
+
+    const headers = [];
+    for (const [key, value] of request.headers) {
+      if (key === 'host') {
+        // By default fetch implementation will join "host" header values with a comma.
+        // But in order to be node.js compatible, we need to select the first if possible.
+        headers.push(key, value.split(', ').at(0) as string);
+      } else {
+        headers.push(key, value);
+      }
+    }
+    incoming._addHeaderLines(headers, headers.length);
+
+    // TODO(soon): It would be useful if there was a way to expose request.cf properties.
```

Comment:
I'd say it's essential, not just useful. We should figure this out before we flip the switch making `node:http` available. Can be just as simple as adding a new own property to `incoming`

Reply:
I'll follow up on this on a different PR.